### ISSUE TITLE
[BUG] printing hashtags in level 4

### ISSUE THREAD

**Issue by OnnoEbbens** [hedyorg/hedy#3882]
For some reason (don't ask me why) a student wanted to print these lines:

```
print 'PPPPPPPPPPPPPP5555555555555555555GP5YYBBBBBBBBBB#BBBBBBBBBBBBBBGP555YYYYYYY5555555555555555555555555'
print 'PPPPPPPPPPPPPPP5555555555555555Y5##BGPG#BBBBBBBBBBBBBBBBBBBBBBBBBBBBBGGPP555555555555555555555555555'
```

This causes an error:

```
We konden je code niet goed lezen.

Let op! Bij print en ask moet voor én achter de tekst een aanhalingsteken komen. Jij bent dat vergeten voor de tekst 'PPPPPPPPPPPPPP5555555555555555555GP5YYBBBBBBBBBB.
```

I think it is because of the hashtags in the code but I am not sure.

**Comment by Felienne** [comment]
Yes! Hashtags create comments amd cause the rest of the line to be ignored. But yeah within quotes that should not be the case!

**Comment by quinnah** [comment]
Hello! I was wondering if this issue is still active / open to work on

**Comment by Felienne** [comment]
> Hello! I was wondering if this issue is still active / open to work on

Yes! It can probably be solved in a similar place as the other issue you were working on! Les us know if you need help.

**Comment by quinnah** [comment]
Just to reproduce the error:
<img width="695" alt="Screenshot 2024-01-09 at 1 54 46 PM" src="https://github.com/hedyorg/hedy/assets/55807397/a0fc94dc-9bdf-4bee-8f63-a2a7fa79e71b">
where we want to display 'My name is #Sophie' rather than the # commenting out the rest of the line.

Would this be an issue in the isValid class where we handle exceptions for parsing different input?

**Comment by Felienne** [comment]
> where we would want to display 'My name is #Sophie' rather than the # commenting out the rest of the line.

Yes, that is what we want!

To be honest, I am also not 100% sure why the code works as it works 😅 But I think your assumptions are valid, the rest of the line is skipped and then so is the closing quote. If you need help diving in, let me know, but I would likely also require a bit of time to look into it (as you can see this issue is old and the code creating it is even older :) )

**Comment by quinnah** [comment]
I did some more digging. Something interesting is that the results are non-deterministic (inconsistent). Here are two consecutive runs with the same input string '#helloworld'
<img width="1423" alt="Screenshot 2024-01-10 at 9 33 47 AM" src="https://github.com/hedyorg/hedy/assets/55807397/c2a19abe-47d7-446d-86c8-1057972ba5af">
<img width="1208" alt="Screenshot 2024-01-10 at 11 18 42 AM" src="https://github.com/hedyorg/hedy/assets/55807397/317b7ae2-e124-4c3e-b234-119e1bdf652d">
The same behavior happens when I test locally, it seems roughly half the time it passes and half the time it doesn't.  


I ran through some Pytest output with the test input string '#Helloworld' to figure out at what point in the program the input text becomes truncated from '#Helloworld' -> ''

As you see below, the output of print_input() returns an AST with the empty string as text (which is incorrect) and error_print_nq.

<img width="1208" alt="Screenshot 2024-01-10 at 10 27 20 AM" src="https://github.com/hedyorg/hedy/assets/55807397/280f9b14-63b9-49f9-83fb-c0398f61b237">



Inside the [parse_input function](https://github.com/hedyorg/hedy/blob/abdecc8ac93b217ab7170349b58921f9c25b5aac/hedy.py#L3420), I added a logging warning for debugging purposes, which outputted the ParseTree result after parser.parse is called on the input string (which is still correctly '#Helloworld' )
`Tree(Token('RULE', 'program'), [Tree(Token('RULE', 'command'), [Tree('error_print_nq', [Tree('text', [Token('__ANON_15', "'")])])])]`
<img width="1207" alt="Screenshot 2024-01-10 at 12 21 40 PM" src="https://github.com/hedyorg/hedy/assets/55807397/d66271ee-9356-4585-9050-23feba4eea47">
<img width="1207" alt="Screenshot 2024-01-10 at 12 18 33 PM" src="https://github.com/hedyorg/hedy/assets/55807397/49e71725-5f69-4513-aa3e-3dfc4b148a3d">


Since the error persists there, this suggests that the issue is with the parse method in the Lark parser. If someone with more context/knowledge of the Lark parser could provide guidance on how to proceed, that would be appreciated - thank you!

**Comment by Felienne** [comment]
Here are some pointers to dive in further:

Here we define comments and ignore them. This ignore is Lark syntax, and I do not know the exact meaning (esp when it happens in the middle of a line in a quote) you could dive into the workings of Lark there:
https://github.com/hedyorg/hedy/blob/abdecc8ac93b217ab7170349b58921f9c25b5aac/grammars/terminals.lark#L1

In level 3 text, we exclude comments from text, and this production is inherited by default in level 4:

[text: /([^\n،,，、#]+)/ -> text](https://github.com/hedyorg/hedy/blob/abdecc8ac93b217ab7170349b58921f9c25b5aac/grammars/level3-Additions.lark#L14C1-L14C1)

What I think we should to (but I am saying this without having tested it, I might be wrong!) is to add a new rule excluding comments in the middle of a line *if and only if* at the end of the line also a quote appears, and put that rule before the text rule. But I would have to dive in deeper to see if this is true, and to be honest, this issue has "low prio" for a reason :) So I hope this helps a bit, I cannot exert all too much effort into it further since we have so many other things on our plates.

Hope this helps though, and again, creating a PR with partial work makes it easier for me to step in, do not be afraid it is wrong!

**Comment by quinnah** [comment]
Thank you for the comments and guidance, it was very helpful!
I've opened a PR with the failing tests.

I tried variations of modifying the grammar. Unfortunately, looking into Lark more, it seems like it's difficult to modify the `%ignore` to handle conditions. Some threads might be relevant (regarding conditional ignores): see [this](https://github.com/lark-parser/lark/issues/896) (from someone who faced a similar issue with the parser)

The `%ignore` works by checking for the expression (in this case a '#') and skipping the rest of the expression. I tried different ways of modify the expression, but that didn't work. My second thought was to implement a rule to 'ignore the ignore' when the condition is met, but this also doesn't work (see: [here](https://github.com/lark-parser/lark/discussions/934) on locally disabling ignore

One possible solution might be to use a[custom lexer](https://lark-parser.readthedocs.io/en/latest/examples/advanced/custom_lexer.html) or Transformer - instead of using `%ignore` to parse comments automatically - but this seemed like a non-trivial amount of work to handle an edge case :( 

I know this issue was marked low-prio, so would love to hear what the team thinks is best! One possible workaround could be to simply add a warning that '#' characters inside quotes is not currently supported in the debug output
